File: sdk/python/kms_paramstore/secret.py
"""The redacting :class:`Secret` value type."""

from __future__ import annotations

__all__ = ["Secret", "REDACTED", "new_secret"]

# What every secret-bearing type prints instead of plaintext.
REDACTED = "[REDACTED]"


class Secret:
    """Holds secret plaintext together with non-sensitive metadata.

    ``Secret`` is deliberately hard to leak: ``repr``, ``str``, and ``format``
    all emit ``"[REDACTED]"``, so it is safe to pass to ``print``, ``logging``,
    f-strings, or ``%``-formatting. Plaintext is only reachable through the
    explicit :meth:`value` / :meth:`string_value` accessors.
    """

    __slots__ = ("_value", "_env", "_app", "_key", "_version", "_content_type")

    def __init__(
        self,
        value: bytes = b"",
        *,
        env: str = "",
        app: str = "",
        key: str = "",
        version: int = 0,
        content_type: str = "",
    ) -> None:
        self._value = bytes(value)
        self._env = env
        self._app = app
        self._key = key
        self._version = version
        self._content_type = content_type

    # --- plaintext accessors (the only way to read the secret) -------------

    @property
    def value(self) -> bytes:
        """The raw secret plaintext as bytes. The only way to read it."""
        return self._value

    @property
    def string_value(self) -> str:
        """The secret plaintext decoded to ``str`` (UTF-8)."""
        return self._value.decode("utf-8")

    # --- non-sensitive metadata --------------------------------------------

    @property
    def env(self) -> str:
        return self._env

    @property
    def app(self) -> str:
        return self._app

    @property
    def key(self) -> str:
        return self._key

    @property
    def namespace(self) -> str:
        return f"{self._env}/{self._app}"

    @property
    def path(self) -> str:
        """Display path ``/env/app/key`` (or the bare key if the namespace is unset)."""
        if self._env and self._app:
            return f"/{self._env}/{self._app}/{self._key}"
        return self._key

    @property
    def version(self) -> int:
        return self._version

    @property
    def content_type(self) -> str:
        return self._content_type

    def is_empty(self) -> bool:
        return len(self._value) == 0

    def clone(self) -> "Secret":
        """Return an independent wrapper with copied plaintext bytes."""
        return Secret(
            bytes(bytearray(self._value)), env=self._env, app=self._app,
            key=self._key, version=self._version, content_type=self._content_type,
        )

    def __len__(self) -> int:
        return len(self._value)

    def __bool__(self) -> bool:
        return len(self._value) > 0

    # --- redaction ----------------------------------------------------------

    def __repr__(self) -> str:  # used by logging, %r, the REPL
        return REDACTED

    def __str__(self) -> str:  # used by print, str(), f-strings
        return REDACTED

    def __format__(self, spec: str) -> str:  # used by format() / f"{x:...}"
        return REDACTED

    # Equality compares plaintext so tests and rotation checks work, but the
    # comparison never renders the value.
    def __eq__(self, other: object) -> bool:
        if isinstance(other, Secret):
            return self._value == other._value
        return NotImplemented

    def __hash__(self) -> int:  # keep hashability disabled to avoid accidental
        raise TypeError("Secret is unhashable to discourage caching by value")


def new_secret(
    value: bytes,
    *,
    env: str = "",
    app: str = "",
    key: str = "",
    version: int = 0,
    content_type: str = "",
) -> Secret:
    """Wrap plaintext in a :class:`Secret` (mainly for tests and tooling)."""
    return Secret(value, env=env, app=app, key=key, version=version, content_type=content_type)

File: sdk/go/kmsclient/secret.go
package kmsclient

import (
	"encoding/json/jsontext"
	"encoding/json/v2"
	"fmt"
	"io"
)

// redactedText is what every secret-bearing type prints instead of plaintext.
const redactedText = "[REDACTED]"

// Secret holds secret plaintext together with non-sensitive metadata. It is
// deliberately hard to leak: its formatting, JSON, and YAML implementations
// all emit "[REDACTED]", so it is safe to pass to fmt.Printf, structured loggers,
// json.Marshal, or yaml.Marshal. Plaintext is only accessible through
// the explicit Value / StringValue accessors.
//
// Secret is a value type; copies redact identically.
type Secret struct {
	// BindKey is a declaration-only credential used by generated configuration
	// stores. Secrets fetched from KMS always leave it empty.
	BindKey string

	value       []byte
	path        string
	version     uint64
	contentType string
}

// NewSecret wraps plaintext in a Secret. It is mainly useful in tests and
// tooling; normal code obtains Secrets from Client.GetSecret.
func NewSecret(value []byte) Secret { return Secret{value: value} }

// Value returns the raw secret plaintext. This is the only way to read it.
func (s Secret) Value() []byte { return s.value }

// StringValue returns the secret plaintext as a string.
func (s Secret) StringValue() string { return string(s.value) }

// Path returns the store path the secret was read from, if known.
func (s Secret) Path() string { return s.path }

// Version returns the secret version that was read, if known.
func (s Secret) Version() uint64 { return s.version }

// ContentType returns the declared content type of the secret, if known.
func (s Secret) ContentType() string { return s.contentType }

// Clone returns an independent copy of the Secret. The plaintext buffer is
// deep-copied while declaration credentials and immutable metadata are
// preserved. Generated stores strip BindKey before retaining resolved values.
func (s Secret) Clone() Secret {
	s.value = append([]byte(nil), s.value...)
	return s
}

// IsZero reports whether the Secret carries no plaintext.
func (s Secret) IsZero() bool { return len(s.value) == 0 }

// String implements fmt.Stringer and always redacts.
func (s Secret) String() string { return redactedText }

// GoString implements fmt.GoStringer (used by the %#v verb) and always redacts.
func (s Secret) GoString() string { return redactedText }

// Format implements fmt.Formatter so that every verb (%v, %s, %+v, %#v, %q, ...)
// redacts the plaintext. %q wraps the redaction in quotes for valid output.
func (s Secret) Format(f fmt.State, verb rune) {
	switch verb {
	case 'q':
		_, _ = fmt.Fprintf(f, "%q", redactedText)
	default:
		_, _ = io.WriteString(f, redactedText)
	}
}

// MarshalJSON always emits the redaction string, so a Secret embedded in a
// JSON-marshaled struct never leaks plaintext.
func (s Secret) MarshalJSON() ([]byte, error) {
	return json.Marshal(redactedText)
}

// MarshalJSONTo is the streaming JSON v2 equivalent of MarshalJSON.
func (s Secret) MarshalJSONTo(out *jsontext.Encoder) error {
	return json.MarshalEncode(out, redactedText)
}

// MarshalYAML redacts both plaintext and declaration credentials in YAML exports.
func (s Secret) MarshalYAML() (any, error) {
	return redactedText, nil
}
